### PR TITLE
holystone series

### PR DESCRIPTION
**Pull request by S-Leader**
`(TITLE)`: type(scope): Brief description *[MOVE TO TITLE]*

Write up a description here

closes issue #XXXX [REMOVE IF IRRELEVANT]

---

I agree to the Contributor License Agreement (CLA).

### [Example Below; REMOVE FOR SUBMISSION]

`(TITLE)` feat(block): Implements Aether Grass Block

Implements the Aether Grass Block class, model, blockstates, language entry, and adds the block to the registry
  
closes issue #1

---

I agree to the Contributor License Agreement (CLA).

### CODE REVIEW

File: src/main/java/com/aetherteam/aether_genesis/data/generators/loot/GenesisBlockLoot.java [hunk 1]
```diff
@@ -26,6 +26,9 @@ public void generate() {
         this.dropSelfDouble(GenesisBlocks.GREEN_AERCLOUD.get());
         this.dropSelfDouble(GenesisBlocks.PURPLE_AERCLOUD.get());
         this.dropSelfDouble(GenesisBlocks.STORM_AERCLOUD.get());
+        this.dropSelfDouble(GenesisBlocks.HOLYSTONE_HEADSTONE.get());
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/data/generators/loot/GenesisBlockLoot.java]: The holystone variants should just use this.dropSelf, they don't have double drops.

**Review by bconlon1 — CHANGES_REQUESTED** [review]
The holystone headstone's model should not require having a duplicate holystone brick texture in the files.

File: src/main/java/com/aetherteam/aether_genesis/data/generators/loot/GenesisBlockLoot.java (hunk 1 above)
Comment (S-Leader) [reply src/main/java/com/aetherteam/aether_genesis/data/generators/loot/GenesisBlockLoot.java]: got it

File: src/main/java/com/aetherteam/aether_genesis/item/combat/PhoenixDartShooterItem.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import net.minecraft.world.entity.projectile.AbstractArrow;
+import net.minecraft.world.item.*;
+import net.minecraft.world.item.enchantment.Enchantment;
+import net.minecraft.world.item.enchantment.Enchantments;
+import net.minecraft.world.level.Level;
+import net.minecraftforge.event.ForgeEventFactory;
+
+import java.util.function.Predicate;
+
+public class PhoenixDartShooterItem extends ProjectileWeaponItem implements Vanishable {
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/item/combat/PhoenixDartShooterItem.java]: I believe you should be able to just extend DartShooterItem. You can use customDart() to create the Phoenix Dart projectile, override getAllSupportedProjectiles to check the dart tag for usable ammo for the dart shooter, and I guess use golden darts as a default dart type parameter.

**Review by bconlon1 — CHANGES_REQUESTED** [review]
If the intention is to have phoenix darts appear as their own new layer, they shouldn't extend GoldenDart and GoldenDartLayer.

**Review by bconlon1 — CHANGES_REQUESTED** [review]
To add onto my previous review comment, that'd also mean the necessary capability and event code to increment the dart stuck layer on the player.

File: src/main/java/com/aetherteam/aether_genesis/item/combat/PhoenixDartShooterItem.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    public ItemStack finishUsingItem(ItemStack stack, Level level, LivingEntity user) {
+        if (user instanceof Player player) {
+            ItemStack ammoItem = player.getProjectile(stack); // Gets matching ammo stack from inventory according to DartShooterItem#getAllSupportedProjectiles().
+                boolean creativeOrShooterIsInfinite = player.getAbilities().instabuild || stack.getEnchantmentLevel(Enchantments.INFINITY_ARROWS) > 0; // Note: Dart shooters can't be enchanted with Infinity in survival, but we still implement the behavior.
+                boolean stillHasAmmo = !ammoItem.isEmpty() || creativeOrShooterIsInfinite;
+
+                ForgeEventFactory.onArrowLoose(stack, level, player, 0, stillHasAmmo);
+
+                if (stillHasAmmo) { // Seems to be a failsafe check; under normal circumstances this should already be true because of the checks in DartShooterItem#use().
+                    if (ammoItem.isEmpty() && stack.is(GenesisTags.Items.DARTS)) {
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/item/combat/PhoenixDartShooterItem.java]: Do you need to check the tag here?

Comment (S-Leader) [reply src/main/java/com/aetherteam/aether_genesis/item/combat/PhoenixDartShooterItem.java]: hmmm, sorry, i'll fix it now

File: src/main/java/com/aetherteam/aether_genesis/capability/player/GenesisPlayerCapability.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package com.aetherteam.aether_genesis.capability.player;
+
+import com.aetherteam.aether.capability.player.AetherPlayerCapability;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.world.entity.player.Player;
+
+public class GenesisPlayerCapability extends AetherPlayerCapability implements GenesisPlayer {
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/capability/player/GenesisPlayerCapability.java]: You shouldn't have to extend AetherPlayerCapability for this, just CapabilitySyncing

File: src/main/java/com/aetherteam/aether_genesis/capability/player/GenesisPlayer.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package com.aetherteam.aether_genesis.capability.player;
+
+import com.aetherteam.aether.capability.player.AetherPlayer;
+import com.aetherteam.aether_genesis.capability.GenesisCapabilities;
+import net.minecraft.world.entity.player.Player;
+import net.minecraftforge.common.util.LazyOptional;
+
+public interface GenesisPlayer extends AetherPlayer {
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/capability/player/GenesisPlayer.java]: You shouldn't have to extend AetherPlayer for this.

File: src/main/java/com/aetherteam/aether_genesis/capability/GenesisCapabilities.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	@SubscribeEvent
+	public static void register(RegisterCapabilitiesEvent event) {
+		event.register(GenesisPlayer.class);
+	}
+
+	@EventBusSubscriber(modid = Genesis.MODID)
+	public static class Registration {
+		@SubscribeEvent
+		public static void attachEntityCapabilities(AttachCapabilitiesEvent<Entity> event) {
+			if (event.getObject() instanceof LivingEntity livingEntity) {
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/capability/GenesisCapabilities.java]: could do `if (event.getObject() instanceof Player player)` instead

File: src/main/java/com/aetherteam/aether_genesis/capability/player/GenesisPlayerCapability.java (hunk 4 above)
Comment (S-Leader) [reply src/main/java/com/aetherteam/aether_genesis/capability/player/GenesisPlayerCapability.java]: got it

File: src/main/java/com/aetherteam/aether_genesis/client/renderer/PhoenixDartLayer.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    public PhoenixDartLayer(EntityRenderDispatcher renderDispatcher, LivingEntityRenderer<T, M> renderer) {
+        super(renderDispatcher, renderer);
+    }
+
+    @Override
+    protected int numStuck(@Nonnull T entity) {
+        if (entity instanceof Player player) {
+            Optional<GenesisPlayer> genesisPlayerOptional = GenesisPlayer.get(player).resolve();
+            if (genesisPlayerOptional.isPresent()) {
+                return genesisPlayerOptional.get().getGoldenDartCount();
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/client/renderer/PhoenixDartLayer.java]: use `.getPhoenixDartCount()`

Comment (S-Leader) [reply src/main/java/com/aetherteam/aether_genesis/client/renderer/PhoenixDartLayer.java]: sorry this is the first time i weite about capability related things, i'm so sorry for do this

File: src/main/java/com/aetherteam/aether_genesis/capability/player/GenesisPlayerCapability.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    public void deserializeSynchableNBT(CompoundTag tag) {
+        if (tag.contains("PhoenixDartCount_Syncing")) {
+            this.setPhoenixDartCount(tag.getInt("PhoenixDartCount_Syncing"));
+        }
+    }
+
+    @Override
+    public AetherPacket getSyncPacket(CompoundTag tag) {
+        return new AetherPlayerSyncPacket(this.getPlayer().getId(), tag);
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/capability/player/GenesisPlayerCapability.java]: needs to be a sync packet specifically for this capability.

File: src/main/java/com/aetherteam/aether_genesis/item/combat/PhoenixDartShooterItem.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+    /**
+     * Based somewhat on {@link net.minecraft.world.item.BowItem#releaseUsing(ItemStack, Level, LivingEntity, int)}, although we use {@link Item#finishUsingItem(ItemStack, Level, LivingEntity)} instead to allow for the Dart Shooter to shoot continuously without releasing.
+     * @param stack The {@link ItemStack} in use.
+     * @param level The {@link Level} of the user.
+     * @param user The {@link LivingEntity} using the stack.
+     * @return The used {@link ItemStack}.
+     */
+    @Override
+    public ItemStack finishUsingItem(ItemStack stack, Level level, LivingEntity user) {
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/item/combat/PhoenixDartShooterItem.java]: you should be able to remove this method.

File: src/main/java/com/aetherteam/aether_genesis/item/combat/PhoenixDartShooterItem.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    public Predicate<ItemStack> getAllSupportedProjectiles() {
+        return (stack) -> stack.is(GenesisTags.Items.DARTS);
+    }
+
+    public AbstractDart customDart(AbstractDart dart) {
+        return dart;
+    }
+
+    public AbstractDart createDart(Level level, LivingEntity shooter) {
```
Comment (bconlon1) [inline src/main/java/com/aetherteam/aether_genesis/item/combat/PhoenixDartShooterItem.java]: move code from this method into customDart and mark customDart as override

**Review by bconlon1 — APPROVED** [review]
lgtm

**Pull request merged**